components/dialog/dialog-mixin.js

```diff
@@ -177,10 +177,33 @@ export const DialogMixin = superclass => class extends RtlMixin(superclass) {
 		}
 	}
 
+	_findAutofocusElement(node) {
+		if (this._useNative) {
+			// Do not override native autofocus attribute implementation
+			return null;
+		}
+
+		if (!node.querySelector('slot')) {
+			// We are in a confirm dialog, autofocus attribute will never be set
+			return null;
+		}
+
+		const content = node.querySelector('slot').assignedElements({ flatten: true });
+		const el = content.find(el => el.autofocus)
+			?? content.find(el => el.querySelector('[autofocus]'))?.querySelector('[autofocus]');
+		return el;
```

Comment:
I had a lot of trouble with this bit. `querySelector('[autofocus]')` will not traverse `slot`s. `getComposedChildren` also did not work.